### ISSUE TITLE
dnf5-5.2.0.0: Downgrading to the dnf5-5.1.* version leaves the system without dnf command

### ISSUE THREAD

**Issue by jan-kolarik** [rpm-software-management/dnf5#1453]
It's probably not an issue when using only packages from the Rawhide repository, as there will be just one version of dnf5 at a time. However, I'd still open the ticket to think about it further.

**Comment by Conan-Kudo** [comment]
It would probably make sense to split the dnf, yum, and microdnf provides into subpackages instead. That neatly resolves this issue.

**Comment by j-mracek** [comment]
I prefer to avoid splitting of dnf, yum, and microdnf provides into subpackages because it would keep the same mess between RHEL, Fedora and containers. Currently DNF is not available in `rhelX-minimal`, `fedora-minimal`, yum is not available in `rhel-minimal`, `fedora` and `fedora-minimal` and `microdnf` is available only in `fedora-minimal` and `rhel-minimal`, and dnf is not available in `fedora-minimal` and `rhel-minimal`. It means there is not a `common` way how to install a package in all of those images. With DNF5 we can provide a unified user experience or we can extend the mess by adding DNF5 binary.

**Comment by ppisar** [comment]
I guess the idea was using empty packages instead of RPM Provides with Obsoletes. That means programs would still be packaged in dnf5 binary package. But empty packages would allow downgrading to, e.g. DNF4, because an identity of the old package manager would not be destroyed with Obsoletes.

However, I agree with the original poster, that this feature is only interested for those who want to support free downgrades. And this is not the case of Fedora as Fedora has decided a move from DNF4 for DNF5. So i personally observe this problem as miniscule from point of Fedoras.

**Comment by evan-goode** [comment]
DNF5 5.1.* is no longer in any stable Fedora, so this issue is no longer relevant.

**Comment by Conan-Kudo** [comment]
FWIW, this wound up being a problem for transitioning Mageia to DNF5 (see [mga#35168](https://bugs.mageia.org/35168)), so I wound up doing a split: 

```diff
--- dnf5.spec.orig	2026-05-21 10:39:35.422775796 -0400
+++ dnf5.spec	2026-05-10 07:49:07.282683197 -0400
@@ -15,9 +15,12 @@
 
 %bcond dnf5_obsoletes_dnf %[0%{?mageia} >= 10]
 
+# For mga#35168
+%bcond dnf_dummy_packages %[0%{?mga10}]
+
 Name:           dnf5
 Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
-Release:        %mkrel 1
+Release:        %mkrel 3
 Summary:        Command-line package manager
 Group:          System/Packaging
 License:        GPL-2.0-or-later
@@ -61,7 +64,7 @@ Provides:       microdnf = %{version}-%{
 Obsoletes:      microdnf < 4
 %endif
 
-%if %{with dnf5_obsoletes_dnf}
+%if %{with dnf5_obsoletes_dnf} && ! %{with dnf_dummy_packages}
 Provides:       dnf = %{version}-%{release}
 Obsoletes:      dnf < 5
 
@@ -422,9 +425,40 @@ It supports RPM packages, modulemd modul
 %exclude %{_datadir}/dnf5/aliases.d/compatibility-reposync.conf
 %endif
 
+# == dummy packages for mga#35168 ==
+
+%if %{with dnf_dummy_packages}
+%package -n dnf
+Summary:        Command-line package manager (transitional package)
+Version:        4.99.0^really%{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
+Requires:       %{name} = %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}-%{release}
+BuildArch:      noarch
+
+%description -n dnf
+This package exists to provide a reasonable upgrade path from
+DNF4 to DNF5 across Mageia releases. It can be safely removed.
+
+%files -n dnf
+# empty file list
+
+%package -n yum
+Summary:        Command-line package manager (transitional package)
+Version:        4.99.0^really%{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
+Requires:       %{name} = %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}-%{release}
+BuildArch:      noarch
+
+%description -n yum
+This package exists to provide a reasonable upgrade path from
+YUM to DNF5 across Mageia releases. It can be safely removed.
+
+%files -n yum
+# empty file list
+%endif
+
 # ========== libdnf5 ==========
 %package -n %{libname}
 Summary:        Package management library
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Libraries
 License:        LGPL-2.1-or-later
 Provides:		libdnf5%{?_isa} = %{version}-%{release}
@@ -488,6 +522,7 @@ Package management library.
 %if %{with libdnf_cli}
 %package -n %{libcliname}
 Summary:        Library for working with a terminal in a command-line package manager
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Libraries
 License:        LGPL-2.1-or-later
 Provides:		libdnf5-cli%{?_isa} = %{version}-%{release}
@@ -507,6 +542,7 @@ Library for working with a terminal in a
 
 %package -n dnf5-devel
 Summary:        Development files for dnf5
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          Development/C++
 License:        LGPL-2.1-or-later
 Requires:       dnf5%{?_isa} = %{version}-%{release}
@@ -526,6 +562,7 @@ Development files for dnf5.
 
 %package -n %{devname}
 Summary:        Development files for libdnf
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          Development/C++
 License:        LGPL-2.1-or-later
 Provides:       libdnf5-devel%{?_isa} = %{version}-%{release}
@@ -548,6 +585,7 @@ Development files for libdnf.
 
 %package -n %{devcliname}
 Summary:        Development files for libdnf5-cli
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          Development/C++
 License:        LGPL-2.1-or-later
 Provides:       libdnf5-cli-devel%{?_isa} = %{version}-%{release}
@@ -569,6 +607,7 @@ Development files for libdnf5-cli.
 %if %{with perl5}
 %package -n perl-libdnf5
 Summary:        Perl 5 bindings for the libdnf library
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          Development/Perl
 License:        LGPL-2.1-or-later
 Requires:       libdnf5%{?_isa} = %{version}-%{release}
@@ -590,6 +629,7 @@ Perl 5 bindings for the libdnf library.
 %if %{with perl5} && %{with libdnf_cli}
 %package -n perl-libdnf5-cli
 Summary:        Perl 5 bindings for the libdnf5-cli library
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          Development/Perl
 License:        LGPL-2.1-or-later
 Requires:       libdnf5-cli%{?_isa} = %{version}-%{release}
@@ -611,6 +651,7 @@ Perl 5 bindings for the libdnf5-cli libr
 %if %{with python3}
 %package -n python3-libdnf5
 Summary:        Python 3 bindings for the libdnf5 library
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          Development/Python
 License:        LGPL-2.1-or-later
 Requires:       libdnf5%{?_isa} = %{version}-%{release}
@@ -632,6 +673,7 @@ Python 3 bindings for the libdnf library
 %if %{with python3} && %{with libdnf_cli}
 %package -n python3-libdnf5-cli
 Summary:        Python 3 bindings for the libdnf5-cli library
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          Development/Python
 License:        LGPL-2.1-or-later
 Requires:       libdnf5-cli%{?_isa} = %{version}-%{release}
@@ -653,6 +695,7 @@ Python 3 bindings for the libdnf5-cli li
 %if %{with ruby}
 %package -n ruby-libdnf5
 Summary:        Ruby bindings for the libdnf library
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          Development/Ruby
 License:        LGPL-2.1-or-later
 Provides:       ruby(libdnf) = %{version}-%{release}
@@ -674,6 +717,7 @@ Ruby bindings for the libdnf library.
 %if %{with ruby} && %{with libdnf_cli}
 %package -n ruby-libdnf5-cli
 Summary:        Ruby bindings for the libdnf5-cli library
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          Development/Ruby
 License:        LGPL-2.1-or-later
 Provides:       ruby(libdnf_cli) = %{version}-%{release}
@@ -694,6 +738,7 @@ Ruby bindings for the libdnf5-cli librar
 %if %{with plugin_actions}
 %package -n libdnf5-plugin-actions
 Summary:        Libdnf5 plugin that allows to run actions (external executables) on hooks
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Packaging
 License:        LGPL-2.1-or-later
 Requires:       libdnf5%{?_isa} = %{version}-%{release}
@@ -716,6 +761,7 @@ Libdnf5 plugin that allows to run action
 
 %package -n libdnf5-plugin-appstream
 Summary:        Libdnf5 plugin to install repository AppStream data
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Packaging
 License:        LGPL-2.1-or-later
 Requires:       libdnf5%{?_isa} = %{version}-%{release}
@@ -735,6 +781,7 @@ Libdnf5 plugin that installs repository'
 %if %{with plugin_expired_pgp_keys}
 %package -n libdnf5-plugin-expired-pgp-keys
 Summary:        Libdnf5 plugin for detecting and removing expired PGP keys
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Packaging
 License:        LGPL-2.1-or-later
 Requires:       libdnf5%{?_isa} = %{version}-%{release}
@@ -756,6 +803,7 @@ Libdnf5 plugin for detecting and removin
 %if %{with plugin_rhsm}
 %package -n libdnf5-plugin-rhsm
 Summary:        Libdnf5 rhsm (Red Hat Subscription Manager) plugin
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Packaging
 License:        LGPL-2.1-or-later
 Requires:       libdnf5%{?_isa} = %{version}-%{release}
@@ -777,6 +825,7 @@ to the subscription levels.
 %if %{with python_plugins_loader}
 %package -n python3-libdnf5-python-plugins-loader
 Summary:        Libdnf5 plugin that allows loading Python plugins
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Packaging
 License:        LGPL-2.1-or-later
 Requires:       libdnf5%{?_isa} = %{version}-%{release}
@@ -799,6 +848,7 @@ Libdnf5 plugin that allows loading Pytho
 %if %{with plugin_local}
 %package -n libdnf5-plugin-local
 Summary:        Libdnf5 plugin that automatically downloads packages to a local repository
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Packaging
 License:        LGPL-2.1-or-later
 Requires:       libdnf5%{?_isa} = %{version}-%{release}
@@ -822,6 +872,7 @@ repository on the local filesystem and g
 %if %{with dnf5daemon_client}
 %package -n dnf5daemon-client
 Summary:        Command-line interface for dnf5daemon-server
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Packaging
 License:        GPL-2.0-or-later
 Requires:       libdnf5%{?_isa} = %{version}-%{release}
@@ -846,6 +897,7 @@ Command-line interface for dnf5daemon-se
 %if %{with dnf5daemon_server}
 %package -n dnf5daemon-server
 Summary:        Package management service with a DBus interface
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Packaging
 License:        GPL-2.0-or-later
 Requires:       libdnf5%{?_isa} = %{version}-%{release}
@@ -888,6 +940,7 @@ Package management service with a DBus i
 
 %package -n dnf5daemon-server-polkit
 Summary:        Polkit rule to allow wheel group members install trusted packages
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Packaging
 License:        GPL-2.0-or-later
 Requires:       polkit
@@ -908,6 +961,7 @@ trusted repositories using dnf5daemon-se
 
 %package -n dnf5-plugins
 Summary:        Plugins for dnf5
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Packaging
 License:        LGPL-2.1-or-later
 Requires:       dnf5%{?_isa} = %{version}-%{release}
@@ -955,6 +1009,7 @@ copr, needs-restarting, repoclosure, rep
 
 %package plugin-automatic
 Summary:        Package manager - automated upgrades
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 Group:          System/Packaging
 License:        LGPL-2.1-or-later
 Requires:       dnf5%{?_isa} = %{version}-%{release}
@@ -997,6 +1052,7 @@ automatically and regularly from systemd
 %if %{with plugin_manifest}
 %package plugin-manifest
 Summary:        DNF5 plugin for working with RPM package manifest files
+Version:        %{project_version_prime}.%{project_version_major}.%{project_version_minor}.%{project_version_micro}
 License:        LGPL-2.1-or-later
 Requires:       dnf5%{?_isa} = %{version}-%{release}
 Requires:       libdnf5%{?_isa} = %{version}-%{release}
```
